Refactor time series

payetvin opened the pull request:
#1669

# TODO
- [X] Wind
- [X] Load #1686
- [X] Solar #1688 
- [x] Thermal #1693 
- [x] Renewable clusters #1693 
- [x] Hydro #1705
- [x] `(year, hour)` or `(hour, year)` ?
- [x] File name for hydro/thermal
- [x] tests
- [ ] Binding constraints ?
- [ ] Links ?
- [ ] use `std::vector` instead of `Matrix<>`
- [ ] ihm

payetvin committed:
[DEV] Add classes TimeSeriesNumbers TimeSeries

payetvin committed:
[DEV] Add classes to cpp

payetvin committed:
[DEV] getTSNumber

payetvin committed:
[DEV] getCoeff getColumn

payetvin committed:
[FIX] Windows CI

payetvin committed:
[DEV] Use matrix<> instead of vector

payetvin committed:
[DEV] Add timeSeriesLoadFromFolder

payetvin committed:
[DEV]
removed pointer from windseries
add savetofolder
add reset
add typedef TSNumbers

payetvin committed:
[DEV] Wind compiling

payetvin committed:
[DEV] remove old wind series files

payetvin committed:
[FIX] crash

payetvin committed:
[FIX] getSeriesIndex

payetvin committed:
[DEV] load series

payetvin committed:
[DEV] remove load series files

payetvin committed:
[FIX] compile

flomnes commented on src/libs/antares/study/parts/common/series.cpp:
```diff
@@ -38,6 +38,75 @@ using namespace Yuni;
+}
+
+int TimeSeries::timeSeriesSaveToFolder(const AreaName& areaID, const std::string& folder,
+                                       const std::string& filename) const
```
~~filename~~ prefix

flomnes commented on src/libs/antares/study/parts/common/series.cpp:
```diff
@@ -38,6 +38,75 @@ using namespace Yuni;
+    buffer.clear() << folder << SEP << filename << areaID << ".txt";
+    ret = timeSeries.saveToCSVFile(buffer, 0) && ret;
+
+    return ret;
```
```suggestion
    buffer.clear() << folder << SEP << filename << areaID << ".txt";
    return timeSeries.saveToCSVFile(buffer, 0);
```

payetvin committed:
[DEV] Comments

flomnes commented on src/libs/antares/study/parts/common/series.h:
```diff
@@ -33,6 +33,51 @@
+class TimeSeries
+{
+public:
+    typedef Matrix<uint32_t> TSNumbers;
```
```suggestion
    using TSNumbers =Matrix<uint32_t>;
```

flomnes replied:
Also add a typedef for `timeSeries`

payetvin committed:
[DEV] revert reset()

payetvin committed:
[DEV] solar series

payetvin committed:
[DEV] remove solar series files

payetvin committed:
[FIX] Tests crashes

payetvin committed:
[DEV] Add static member emptyColumn

payetvin committed:
[DEV] remove useless if

payetvin committed:
Use generic class for load series (#1686)

* [DEV] load series

* [DEV] remove load series files

* [FIX] compile

* [DEV] Comments

* [DEV] revert reset()

payetvin committed:
Merge branch 'feature/series' of https://github.com/AntaresSimulatorTeam/Antares_Simulator into feature/series

payetvin committed:
Merge branch 'feature/series-solar' into feature/series

payetvin committed:
[DEV] Comments

payetvin committed:
[DEV] using TScoefficients

flomnes commented on src/libs/antares/study/parts/common/series.h:
```diff
@@ -33,6 +33,54 @@
+    uint64_t memoryUsage() const;
+
+    TScoefficients timeSeries;
+    TSNumbers timeseriesNumbers;
```
In some cases (thermal, hydro), we have multiple `TScoefficients` objects sharing a single instance of TSNumbers. For example availability, CO2 cost, fuel cost, etc.

```suggestion
    TScoefficients timeSeries;
    const TSNumbers& timeseriesNumbers;
```

sylvlecl commented on src/libs/antares/study/parts/common/series.h:
```diff
@@ -33,6 +33,54 @@
 namespace Antares::Data
 {
+
+class TimeSeries
```
We should take the opportunity of this refactoring to have some unit tests on those classes, in particular this one.

We need to test the indexing logic and the serialization methods.

payetvin committed:
Refactor thermal & renewable series (#1693)

* [DEV] Thermal series start

* [FIX] getCoefficient() arg order

* [DEV] Use Matrix<double> only

* [DEV] Remove class DataSeriesCommon

* [DEV] Add using TS

* [skip ci] Add library series, move files to part/series

* [DEV] Use library series

* [DEV] Replace matrix<double> with TimeSeries::TS

* [DEV] Remove template from class TimeSeriesConfigurer

* [DEV] Clean valueAtTimeStep

payetvin committed:
[DEV] Comments and use series.reset()

payetvin committed:
Refactor hydro series (#1705)

* [DEV] Removed hydro/series.hxx

* [DEV] Use a ref for TSNumbers

* [DEV] Use a ref for TSNumbers (#1706)

* [DEV] Use TimeSeries for ror

* [DEV] Modify operator[] behavior

* [DEV] storage

* [CI] Upgrade python to 3.12

* [DEV] mingen

* [DEV] replace tsindex with year in hydro/management

* [DEV] Remove getIndex from hydro series

* [DEV] code smells

* Make hydro ts count a private member (#1712)

* Refactor max power TS checks : internalize count of generation time series (ror, storage, mingen)

We should not be able to change a TS count without resizing the TS.
This was allowed when because this count was a public data member.
So we make it a private data member, and change it only when resizing the associated data.

* Make hydro time series cont a private member

---------

Co-authored-by: guilpier-code <62292552+guilpier-code@users.noreply.github.com>

payetvin committed:
[MERGE] Develop + conflicts in management and utils

payetvin committed:
[FIX] Compile

payetvin committed:
[FIX] Code smells

payetvin committed:
[FIX] Code smells

sylvlecl commented on src/libs/antares/study/CMakeLists.txt:
```diff
@@ -1,3 +1,21 @@
+
+target_include_directories(series
+        PUBLIC
+        ${CMAKE_CURRENT_SOURCE_DIR}/parts/series
```
We should move this library to the same level as other libraries: array, study, etc. This CMakeLists.txt should be only dedicated to the study lib, to make things clear.
Here it's really not clear: all other subfolders are part of the "study" library, but only this subfolder is a separate library.

Also we must not expose this directory as a public header directory: instead only the new "include" directory must be defined as the public header directory.

sylvlecl commented on src/libs/antares/study/parts/series/include/antares/series/series.h:
```diff
@@ -0,0 +1,90 @@
+{
+public:
+    using TSNumbers = Matrix<uint32_t>;
+    using TS = Matrix<double>;
```
It seems a little dangerous to remove the fact that some classes were using integer types as the read/write type: what are the impacts, in particular on the UI ?

I'll try to think a little more about this.

sylvlecl replied:
Here again there's a redundancy with encompassing `TimeSeries`, maybe use `Values` ?

payetvin committed:
[DEV] valueAtTimeStep uses year hour now

payetvin committed:
Unit tests for timeseries (#1713)

sylvlecl commented on src/libs/antares/study/parts/hydro/series.h:
```diff
@@ -27,6 +27,7 @@
 #ifndef __ANTARES_LIBS_STUDY_PARTS_HYDRO_TIMESERIES_H__
 #define __ANTARES_LIBS_STUDY_PARTS_HYDRO_TIMESERIES_H__
 
+#include <include/antares/series/series.h>
```
We should not have the "include" folder, here, see comment on the définition of the include dir in CMakeLists.txt

sylvlecl reviewed (changes requested):
First comments just about the new lib definition:
we need to move it outside the study folder, and to fix its include directory.

sylvlecl commented on src/libs/antares/study/parts/thermal/ecoInput.h:
```diff
@@ -86,8 +87,8 @@ class EconomicInputData
 
     //! All {FO,PO}{Duration,Rate} annual values
     // max x DAYS_PER_YEAR
-    Matrix<double> fuelcost;
```
Not sure it's relevant to use that type here since it was not really involved in this work.

sylvlecl commented on src/libs/antares/study/parts/series/include/antares/series/series.h:
```diff
@@ -0,0 +1,90 @@
+     ** \param filename The filename of the series
+     ** \return A non-zero value if the operation succeeded, 0 otherwise
+     */
+    bool timeSeriesLoadFromFolder(const std::string& path,
```
Just `loadFromFolder` to avoid name redundancy

sylvlecl commented on src/libs/antares/study/parts/series/include/antares/series/series.h:
```diff
@@ -0,0 +1,90 @@
+     ** \param filename The filename of the series
+     ** \return A non-zero value if the operation succeeded, 0 otherwise
+     */
+    int timeSeriesSaveToFolder(const AreaName& areaID, const std::string& folder,
```
Just `saveToFolder` to avoid name redundancy

sylvlecl commented on src/libs/antares/study/parts/series/include/antares/series/series.h:
```diff
@@ -0,0 +1,90 @@
+class TimeSeries
+{
+public:
+    using TSNumbers = Matrix<uint32_t>;
```
Since it's already in `TimeSeries` prefix, we could rename it just `Numbers`.

sylvlecl commented on src/libs/antares/study/parts/series/include/antares/series/series.h:
```diff
@@ -0,0 +1,90 @@
+namespace Antares::Data
+{
+
+class TimeSeries
```
We should add some documentation comment to clarify the goal of this class (encapsulating the redirection to the correct timeseries for a given year, basically)

sylvlecl commented on src/libs/antares/study/parts/series/include/antares/series/series.h:
```diff
@@ -0,0 +1,90 @@
+    void reset();
+    void unloadFromMemory() const;
+    void roundAllEntries();
+    void resize(uint32_t year, uint32_t hour);
```
Plural `years` and `hours` would be clearer.

But actually "year" is misleading here: usually what we call years are the monte carlo years. But here, we are not changing the number of MC years, we are changing the number of timeseries. So we could call the parameter "timeSeriesCount" ?

sylvlecl commented on src/libs/antares/study/parts/series/include/antares/series/series.h:
```diff
@@ -0,0 +1,90 @@
+    const double* getColumn(uint32_t year) const;
+    uint32_t getSeriesIndex(uint32_t year) const;
+
+    double* operator[](uint32_t index);
```
Thanks for renaming it to index, just a thought:
it makes me think that it was probably not a great idea to mix in the same class the access by index and the access by years.
We will really need further refactoring after this PR to separate better those 2 concerns.

sylvlecl commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -78,84 +94,87 @@ bool DataSeriesHydro::loadFromFolder(Study& study, const AreaName& areaID, const
-            logs.error() << "Hydro: `" << areaID
-                         << "`: empty matrix detected. Fixing it with default values";
-            ror.reset(1, HOURS_PER_YEAR);
-            storage.reset(1, DAYS_PER_YEAR);
```
Here we have changed the behaviour: we had days per year.

It's an issue, it means that the index in `TimeSeries` class is not always an hour, so we cannot name it hour in that class.

payetvin committed:
[DEV] Fix include path

payetvin committed:
Merge branch 'feature/series' of https://github.com/AntaresSimulatorTeam/Antares_Simulator into feature/series

sylvlecl commented on src/libs/antares/study/study.h:
```diff
@@ -610,7 +610,7 @@ class Study: public Yuni::NonCopyable<Study>, public IObject, public LayerData
     ScenarioBuilder::Sets* scenarioRules = nullptr;
     //@}
 
-    Matrix<double> scenarioHydroLevels;
```
Here too, not sure it's relevant to change the type for now.

sylvlecl commented on src/solver/hydro/management/management.cpp:
```diff
@@ -328,7 +320,7 @@ bool HydroManagement::checkHourlyMinGeneration(uint tsIndex, Data::Area& area) c
                         << "In area: " << area.name << " [hourly] minimum generation of "
                         << srcmingen[day * 24 + h] << " MW in timestep " << day * 24 + h + 1
-                        << " of TS-" << tsIndex + 1
+                        << " of TS-" << year + 1
```
We either need to change the log, or to retrieve the corresponding TS index here (unfortunately). Better to get the TS index, so the user understands what data it is.
Otherwise it's misleading: when reading the log, we will understand it's the TS number, whereas it's actually the MC year.

sylvlecl commented on src/solver/hydro/management/management.cpp:
```diff
@@ -245,34 +239,32 @@ bool HydroManagement::checkMonthlyMinGeneration(uint numSpace, uint tsIndex, con
         // Yearly minimum generation <= Yearly inflows
         logs.error() << "In Area " << area.name << " the minimum generation of "
-                     << data.totalYearMingen << " MW of TS-" << tsIndex + 1
+                     << data.totalYearMingen << " MW of TS-" << year + 1
```
Same, TS index

sylvlecl commented on src/solver/hydro/management/management.cpp:
```diff
@@ -245,34 +239,32 @@ bool HydroManagement::checkMonthlyMinGeneration(uint numSpace, uint tsIndex, con
             logs.error() << "In Area " << area.name << " the minimum generation of "
                          << data.totalMonthMingen[realmonth] << " MW in month " << month + 1
-                         << " of TS-" << tsIndex + 1 << " is incompatible with the inflows of "
+                         << " of TS-" << year + 1 << " is incompatible with the inflows of "
```
Same, TS index

sylvlecl commented on src/solver/hydro/management/management.cpp:
```diff
@@ -234,7 +228,7 @@ void HydroManagement::minGenerationScaling(uint numSpace, uint year) const
 }
 
-bool HydroManagement::checkMonthlyMinGeneration(uint numSpace, uint tsIndex, const Data::Area& area) const
+bool HydroManagement::checkMonthlyMinGeneration(uint numSpace, uint year, const Data::Area& area) const
```
we need to change the parameter name in headers too, same for all methods

sylvlecl commented on src/solver/simulation/common-hydro-remix.cpp:
```diff
@@ -96,8 +96,8 @@ static bool Remix(const Data::AreaList& areas, PROBLEME_HEBDO& problem, uint num
+            auto& load = area.load.series.timeSeries;
             assert(load.width > 0);
 
             auto& L = (loadTS < load.width) ? load[loadTS] : load[0];
```
should be replaced with call to getCOlumn, and remove the getSeriesIndex above

sylvlecl commented on src/solver/simulation/sim_calcul_economique.cpp:
```diff
@@ -574,32 +574,31 @@ void SIM_RenseignementProblemeHebdo(const Study& study,
+            auto loadSeries = area.load.series.getCoefficient(year, hourInYear);
+            auto windSeries = area.wind.series.getCoefficient(year, hourInYear);
+            auto solarSeries = area.solar.series.getCoefficient(year, hourInYear);
+            auto rorSeriesIndex = area.hydro.series->ror.getSeriesIndex(year);
```
I think we can just get the rorSeries like for others, not the index

sylvlecl reviewed (dismissed):
Some more comments.

A quite important one is that not all timeseries are actually hourly, at least one is a daily one.
So we need to make a choice:
- either use the same class and rename the "hour" parameters in a more generic "timestep"
- or separate between 2 classes

The benefit of the second one is to document by the type what kind of data is actually expected, which would be nice.
In the "cons", it would probably imply some code duplication. Or just some template parameter, but if we do this we lose the benefit of being able to name correctly the arguments (`hour` and `day`).

payetvin committed:
[DEV] move folder series

payetvin committed:
[DEV] Use timestep instead of hour

payetvin committed:
[FIX] several comments for class series

payetvin committed:
[FIX] more comments

payetvin committed:
[FIX] try fix windows CI
